Allow the photometry aperture to be a multiple of the FWHM in each image

mwcraig opened the pull request:
The title says most of it, but a couple of points that will not be obvious from the code itself that should be considered:

+ The "gap" and "width" of the annulus are currently still in pixels even though the "radius" is a multiplier of the FWHM iv `variable_aperture` is True. So, for example, if `variable_aperture` is `True`, and the fwhm is 5 and the aperture radius is 1.5, the radius in pixels is 7.5. **The key question here is whether gap and annulus should be pixels or something else.**
+ I'm testing all of the new code, but the variable aperture test only checks that the FWHM that was put into the three generated test images is correct, and that the correct aperture radius is used. I think that is fine, since aperture photometry itself is checked elsewhere.

mwcraig commented:
Not sure why codecov is so far off on coverage (it is 100% in source detection), but rebased to see what happens.

JuanCab commented on stellarphot/gui_tools/seeing_profile_functions.py:
```diff
@@ -574,6 +574,8 @@ def _update_plots(self):
         if self.rad_prof is None:
             return
 
+        fwhm = 2 * self.rad_prof.HWHM
```
So here you are setting the FWHM to 2*HWHM, what did you use previously?  Or I guess, based on line 608 that you are replacing putting the cutoff radius as a vertical line in the plot and making it FWHM?   Might want to clarify this in comments.

mwcraig replied:
There are two places below that need the FWHM. It turns out that `rad_prof` has a `FWHM` attribute, so now I just use that in the two places that the FWHM is needed.

JuanCab commented on stellarphot/photometry/photometry.py:
```diff
@@ -380,7 +393,11 @@ def single_image_photometry(
         try:
             xcen, ycen = centroid_sources(
-                ccd_image.data, xs, ys, box_size=2 * photometry_apertures.radius + 1
+                # TODO: FIX THIS
```
Is this a reminder that you need finish something here?

mwcraig replied:
I think it was just a reminder to change the aperture radius to be in pixels, i.e. to use `radius_pixel` instead of `radius`.

mwcraig replied:
Removed it.

JuanCab commented on stellarphot/photometry/photometry.py:
```diff
@@ -380,7 +393,11 @@ def single_image_photometry(
         try:
             xcen, ycen = centroid_sources(
-                ccd_image.data, xs, ys, box_size=2 * photometry_apertures.radius + 1
+                # TODO: FIX THIS
+                ccd_image.data,
+                xs,
+                ys,
```
Funny, the AI caught this as well...

mwcraig replied:
I find its summaries _really_ nice.

JuanCab commented on stellarphot/photometry/source_detection.py:
```diff
@@ -361,3 +364,178 @@ def source_detection(
+
+    with warnings.catch_warnings():
+        # block_reduce generates some warnings about things like unit, wcs, etc
+        # that are set on ccd buut not preserved in the reduced image.
```
Typo
```suggestion
        # that are set on ccd but not preserved in the reduced image.
```

JuanCab commented on stellarphot/photometry/source_detection.py:
```diff
@@ -361,3 +364,178 @@ def source_detection(
+
+    # Only keep the n_brightest sources (eventually)
+    n_brightest = min(len(block_sources), int(n_brightest_sources))
+    # Pad n_bnrightest a bit in case some of these end up being too bright
```
```suggestion
    # Pad n_brightest a bit in case some of these end up being too bright

JuanCab commented on stellarphot/photometry/source_detection.py:
I think I understood most of what you were doing here, looks reasonable.

JuanCab commented on stellarphot/photometry/tests/fake_image.py:
```diff
@@ -24,10 +25,46 @@ class FakeImage:
         the seed will be randomly generated.
```
For consistency, please describe ALL the parameters including the new fwhm and n_repeats in the docstring.

mwcraig replied:
Fixed,

JuanCab commented on stellarphot/photometry/tests/fake_image.py:
```diff
@@ -24,10 +25,46 @@ class FakeImage:
+            # Reshape to n_repeats x n_repeats
+            amplitude_scales = amplitude_scales.reshape((n_repeats, n_repeats))
+            extra_sources = []
+            for i in range(n_repeats):
```
This nested pair of loops seems to suggest you apprend `n_repeats` SQUARED sources to `extra_sources`.  That doesn't seem to make sense given the parameter  name.  Can you clarify what this nested loop is supposed to do.

mwcraig replied:
It is supposed to produced `n__repeats**2` copies. I'll come up with a better name when I document the arguments.

mwcraig replied:
Renamed the variable to `n_repeats_per_side`

JuanCab reviewed (commented):
See the comments, looks mostly good, but could use some additional commenting.

JuanCab commented:
> * The "gap" and "width" of the annulus are currently still in pixels even though the "radius" is a multiplier of the FWHM iv `variable_aperture` is True. So, for example, if `variable_aperture` is `True`, and the fwhm is 5 and the aperture radius is 1.5, the radius in pixels is 7.5. **The key question here is whether gap and annulus should be pixels or something else.**

What is the "something else" you are curious about, as in should they be in multiples of FWHM?

If I am recalling correctly, the outer annulus is used for background estimation, so in that regard, we just need to get far enough out that very little of the star's light is out there contributing to the background estimate.  Is 1.5*FWHM far enough out to avoid enough "starlight" contributing to the counts in the outer annulus?

If I am misinterpreting the concern, please clarify.

mwcraig commented:
@JuanCab -- I think I've addressed all of the review comments. Thanks for the quick review!

> What is the "something else" you are curious about, as in should they be in multiples of FWHM?

I didn't have anything super concrete in mind. I could see leaving `gap` and `width` in pixels. I could also see setting gap to some somewhat bigger multiple of the FWHM (so if radius is 1.5x FWHM, then gap is 2x FWHM, say) and setting the width so that the number of annulus pixels is much larger than the number of aperture pixels.

Leaving it as is is fine for now I think.

> Is 1.5*FWHM far enough out to avoid enough "starlight" contributing to the counts in the outer annulus?

Not really -- I'd expect it is better to use, e.g. 2x FWHM, so that the gap is 0.5x FWHM.

JuanCab reviewed (approved):
Looks good.
